doc/check_ldap.1
'\" t
.\"     Title: CHECK_LDAP
.\"    Author: Marius Rieder <marius.rieder@durchmesser.ch>
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 06/02/2014
.\"    Manual: Monitoringplug  Manual
.\"    Source: check_ldap
.\"  Language: English
.\"
.TH "CHECK_LDAP" "1" "06/02/2014" "check_ldap" "Monitoringplug Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
check_ldap \- Check OpenLDAP Monitoring Database\&.
.SH "SYNOPSIS"
.HP \w'\fBcheck_ldap\fR\ 'u
\fBcheck_ldap\fR [\fB\-\-uri\ \fR\fB\fILDAP\-URI\fR\fR]
.SH "DESCRIPTION"
.PP
Check the status of OpenLDAP slapd by query the monitor database\&.
.SH "OPTIONS"
.PP
The Monitoringplug default options
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Print detailed help screen\&.
.RE
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 4
Print version information\&.
.RE
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Show details for command\-line debugging\&.
.RE
.PP
\fB\-t\fR, \fB\-\-timeout=\fR\fB\fIINTEGER\fR\fR
.RS 4
Seconds before the check return with a timeout alert\&.
.RE
.PP
\fB\-\-eopt=\fR\fB[\fIsection\fR]\fR\fB[@\fIfile\fR]\fR
.RS 4
Read extended options from file\&.
.RE
.PP
\fB\-\-perfdata\fR
.RS 4
Print performance data (if available)\&.
.RE
.PP
Check specific options
.PP
\fB\-H\fR, \fB\-\-uri=\fR\fB\fILDAP\-URI\fR\fR
.RS 4
The LDAP\-URI to connect to\&.
.RE
.PP
\fB\-D\fR, \fB\-\-binddb=\fR\fB\fIBINDDB\fR\fR
.RS 4
The DN to use for binding to the LDAP\&.
.RE
.PP
\fB\-W\fR, \fB\-\-password=\fR\fB\fIPASSWORD\fR\fR
.RS 4
The Password to use for bindung\&. to the LDAP\&.
.RE
.PP
\fB\-w\fR, \fB\-\-warning=\fR\fB\fIDURATION\fR\fR
.RS 4
Return warning if interaction expire time limit\&. (Default to 2s)
.RE
.PP
\fB\-c\fR, \fB\-\-critical=\fR\fB\fIDURATION\fR\fR
.RS 4
Return critical if interaction expire time limit\&. (Default to 5s)
.RE
.SH "SEE ALSO"
.PP
\fBmonitoringplug-eopt\fR(5),\fBmonitoringplug-range\fR(5)
.SH "AUTHOR"
.PP
\fBMarius Rieder\fR <\&marius\&.rieder@durchmesser\&.ch\&>
.RS 4
For monitoringplug
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2012 Marius Rieder
.br
.PP
This manual page was written for Monitoringplug (and may be used by others)\&.
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 or (at your option) any later version published by the Free Software Foundation\&.
.sp


debian/changelog
monitoringplug (0.17) UNRELEASED; urgency=low

  * Rework autotools build system.
  * Allow % sign in ranges and parse them.
  * Use hiredis library for redis checks.
  * Use long for redis memory size.
  * Add check_redis_slave check.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Tue, 16 Oct 2013 20:12:29 +0200

monitoringplug (0.16.1) monitoringplug; urgency=low

  * Make .deb more interoperable

 -- Marius Rieder <marie@pcmarie.nine.ch>  Tue, 15 Oct 2013 17:34:17 +0200

monitoringplug (0.16) monitoringplug; urgency=low

  * Added PostgreSQL checks
  * Added FastCGI checks
  * Added Varnish check
  * Improve Redis check
  * Added SNI support to check_ssl_cert
  * Some bugfixes.


 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sun, 08 Sep 2013 09:43:46 +0200


monitoringplug (0.15) monitoringplug; urgency=low

  * Added check_redis

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sun, 02 Jun 2013 13:24:39 +0200

monitoringplug (0.14) monitoringplug; urgency=low

  * Added check_oping
  * Added check_dns_authoritative      
  * Hidden master support for check_dns_sync      
  * Added check_memcached 
  * Fix some minor bugs

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sat, 01 Dec 2012 18:04:22 +0200

monitoringplug (0.13) monitoringplug; urgency=low

  * Added check_snmp_usv check (Credits to Oliver Schonefeld)
  * Some snmp bugfixed
  * check_ipmi_*: Ignore unreadable sensors.
  * Build system improvments.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Tue, 31 Jul 2012 11:55:23 +0200

monitoringplug (0.12) monitoringplug; urgency=low

  * Fix a SNMP, eopt, perfdata and clustat bug.
  * Improved check_akcp.
  * Added STARTTLS support to check_ssl_cert

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sun, 23 Jul 2012 17:55:23 +0200

monitoringplug (0.11) monitoringplug; urgency=low

  * Added IPMI checks with OpenIPMI.
  * Added GSM-Modem signal quality check.
  * Added SMS and mail notification command.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Sat, 14 Jul 2012 16:51:23 +0200

monitoringplug (0.10) monitoringplug; urgency=low

  * Initial DEB Release.

 -- Marius Rieder <marius.rieder@durchmesser.ch>  Tue, 22 May 2012 20:06:46 +0200


ldap/check_ldap.c
/***
 * Monitoring Plugin - check_ldap.c
 **
 *
 * check_ldap - Check a LDAP Server and fetch statistics.
 *
 * Copyright (C) 2014 Marius Rieder <marius.rieder@durchmesser.ch>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * $Id$
 */

const char *progname  = "check_ldap";
const char *progdesc  = "Check a LDAP Server and fetch statistics.";
const char *progvers  = "0.1";
const char *progcopy  = "2014";
const char *progauth  = "Marius Rieder <marius.rieder@durchmesser.ch>";
const char *progusage = "-H <LDAP-URI> [-b <binddn>] [-w <bindpw>]";

/* MP Includes */
#include "mp_common.h"
#include "ldap_utils.h"
/* Default Includes */
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ldap.h>

/* Global Vars */
thresholds *time_thresholds = NULL;

int main (int argc, char **argv) {
    /* Local Vars */
    struct timeval start_time;
    double time_delta;

    /* Set signal handling and alarm */
    if (signal(SIGALRM, timeout_alarm_handler) == SIG_ERR)
        critical("Setup SIGALRM trap failed!");

    /* Process check arguments */
    if (process_arguments(argc, argv) != OK)
        unknown("Parsing arguments failed!");

    /* Start plugin timeout */
    alarm(mp_timeout);


    /* Start LDAP connection */
    gettimeofday(&start_time, NULL);
    LDAP *ld = NULL;
    ld = mp_ldap_init(NULL);
    LDAPMessage *msg = NULL;

    msg = mp_ldap_search(ld, "cn=Statistics,cn=Monitor", LDAP_SCOPE_ONE,
            NULL,
            (char *[]){"cn", "monitorCounter", '\0'});

    time_delta = mp_time_delta(start_time);

    LDAPMessage *stat = NULL;
    for (stat = ldap_first_entry(ld, msg); stat != NULL;
            stat = ldap_next_entry(ld, stat)) {
        char *cn = NULL;
        char *counter = NULL;
        struct berval **vals;
    
        vals = ldap_get_values_len(ld, stat, "cn");
        if (vals == NULL) {
            if (mp_verbose > 0)
                printf("Skipping %s\n", ldap_get_dn(ld, stat));
            continue;
        }
        cn = strdup(vals[0]->bv_val);
        ldap_value_free_len(vals);
    
        vals = ldap_get_values_len(ld, stat, "monitorCounter");
        if (vals == NULL) {
            if (mp_verbose > 0)
                printf("Skipping %s\n", ldap_get_dn(ld, stat));
            free(cn);
            continue;
        }
        counter = strdup(vals[0]->bv_val);
        ldap_value_free_len(vals);
    
        if (mp_verbose > 0) {
            printf("%s: %s\n", cn, counter);
        }
    
        // Calculate slave lag
        mp_perfdata_int(cn, (int)atol(counter), "c", NULL);
    
        free(cn);
        free(counter);
    }

    mp_perfdata_float("time", (float)time_delta, "s", time_thresholds);

    switch(get_status(time_delta, time_thresholds)) {
        case STATE_OK:
            free_threshold(time_thresholds);
            ok("LDAP");
            break;
        case STATE_WARNING:
            free_threshold(time_thresholds);
            warning("LDAP is slow.");
            break;
        case STATE_CRITICAL:
            free_threshold(time_thresholds);
            critical("LDAP is real slow.");
            break;
    }
    free_threshold(time_thresholds);

    critical("You should never reach this point.");
}

int process_arguments (int argc, char **argv) {
    int c;
    int option = 0;

    static struct option longopts[] = {
        MP_LONGOPTS_DEFAULT,
        LDAP_LONGOPTS,
        MP_LONGOPTS_END
    };

    /* Set default */
    setWarnTime(&time_thresholds, "2s");
    setCritTime(&time_thresholds, "5s");

    while (1) {
        c = mp_getopt(&argc, &argv, MP_OPTSTR_DEFAULT"w:c:"LDAP_OPTSTR, longopts, &option);

        if (c == -1 || c == EOF)
            break;

        getopt_ldap(c);
        getopt_wc_time(c, optarg, &time_thresholds);
    }

    /* Check requirements */
    if (!mp_ldap_uri)
        usage("LDAP-URI is mandatory.");

    return(OK);
}

void print_help (void) {
    print_revision();
    print_revision_ldap();
    print_copyright();

    printf("\n");

    printf("Check description: %s", progdesc);

    printf("\n\n");

    print_usage();

    print_help_default();
    print_help_ldap();
    print_help_warn_time("2s");
    print_help_crit_time("5s");
}

/* vim: set ts=4 sw=4 et syn=c : */
